### DISCUSSION TITLE
k8s 内多套环境使用-同域名-不同ns 如何进行部署

### DISCUSSION THREAD

**Discussion by rocklinux9527** [apache/apisix#10028] [Q&A]
1.背景

       1.公司内网有多套环境 分别使用 ns 做为环境标识。每一个环境 都具有相同域名  分别转发给给同ns 间svc进行访问.
 
  
1.如何使用apisix 简化环境部署方案？
1.思路一方案可行性？
     每套环境部署一个apisix,每个环境中有相同的域名，分别指向不同环境，如何在创建ApisixRoute的时候选择不同的apisix呢？

**Comment by Ethan-Zhang-ZH** [comment]
不同环境都是一样的域名？那平常是怎么测试的？

**Comment by rocklinux9527** [comment]
1. 通过HTTP cookie 或者 HTTP请求参数来转发到不同的  环境相同的域名 (统一使用 OpenResty来处理转发)  
2.跨集群调用使用 绑定 hosts 方式.

**Reply by Ethan-Zhang-ZH** [reply]
觉得没必要多套apisix，多条路由，然后用高级匹配就行了

**Comment by rocklinux9527** [comment]
计划 apisix 部署在K8s 集群中 目标 通过命名空间来区分环境 由于每一个相同的域名 多条路由可以实现这个方案吗？部署多套apisix控制器 为了解决配置ingress 相同域名问题


---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 发送日期 | 2023年10月10日 10:11 |
| 收件人 | apache/apisix ***@***.***> |
| 抄送人 | rocklinux9527 ***@***.***>,
Author ***@***.***> |
| 主题 | Re: [apache/apisix] k8s 内多套环境使用-同域名-不同ns 如何进行部署 (Discussion #10028) |

觉得没必要多套apisix，多条路由，然后用高级匹配就行了

**Comment by flylan** [comment]
哈哈，居然遇到一个跟我公司之前项目使用一样的部署方式的问题。我现在项目以前也是所有环境共用一个域名，生产环境，预发布环境，测试环境，通过cookie方式切换到不同的环境去。

好处就是方便，产品和测试通过切换cookie，可以很方便切换不同的环境验收功能。

但是坏处也很明显。
1.本地开发环境和测试环境要做好域名挟持，小心有大聪明在调试测试环境脚本时，那个脚本需要请求一个内部接口的时候，跑到线上环境去触发线上脚本去。哈哈
2.由于多环境共用域名，有时可能某个需求需要运维去改动域名配置，这个需求往往只是处于测试环境，但是发现域名配置要在生产环境去改，最后运维不小心把线上改挂了。
3.多环境共用域名，测试有时遇到问题，压根不知道自己处于生产环境还是测试环境，然后经常误报。

最后我接手项目的时候，强行把测试环境域名和生产域名分离了。不过现在生产环境和预发布环境还是共用一个域名。

**Comment by flylan** [comment]
你的需求可能要通过apisix的插件来实现，根据cookie动态转发到不同的k8s的svc内网域名去。你把生产和测试环境部署在不同的ns下，它们都有一个独立的svc域名。apisix可以通过插件转发过去。不知道现有插件是否能实现，我印象中现在是有个插件支持cookie分流的，如果不满足你的使用场景，可能就需要自己实现一个apisix插件了。
